File: SabreCG_PY/Aircraft.py
from Station import Station
from Leg import Leg
from typing import List

class Aircraft:
    _count = 0
    """
    _tail: Tail Number
    _id
    _count
    _startT
    _endT
    _depStation
    _arrStation
    _dual
    _planLegList
    """
    def __init__(self, name: str, startT: float, endT: float, depS: Station, arrS: Station):
        self._planLegList = []
        self._tail = name
        self._startT, self._endT = startT, endT
        self._depStation, self._arrStation = depS, arrS
        self._id = _count
        _count += 1
        self._dual = 0
    
    def print(self) -> None:
        startT, endT = self._startT, self._endT
        print("Tail %s" % self._tail)
        print("%s %s" % (self._depStation.getName(), startT))
        print("%s %s" % (self._arrStation.getName(), endT))

    def getTail(self) -> str:
        return self._tail
    
    def pushPlanLeg(self, leg: Leg) -> None:
        self._planLegList.append(leg)

    def getDepStation(self) -> Station:
        return self._depStation
    
    def getArrStation(self) -> Station:
        return self._arrStation
    
    def getStartTime(self) -> float:
        return self._startT
    
    def getEndTime(self) -> float:
        return self._endT
    
    def getId(self) -> int:
        return self._id
    
    def setDual(self, dual: float) -> None:
        self._dual = dual

    def getDual(self) -> float:
        return self._dual
    
    def getPlanLegList(self) -> List[Leg]:
        return self._planLegList

    def sortScheLegByDepTime(self):
        self._planLegList.sort(key = lambda _leg: _leg.compareDepKey)

    def isPlanLegFeasible(self):
        pass

File: SabreCG_multipara/SabreCG/Aircraft.cpp
#include"Aircraft.h"
//* #include"Lof.h"
//* #include"AircraftNode.h"

Aircraft::Aircraft(string name, time_t startT, time_t endT, Station * depS, Station * arrS)
	:_tail(name),_depStation(depS),_arrStation(arrS), _startT(startT), _endT(endT)
{
	_id = _count;
	_count++;
	_dual = 0;	//* 对应master problem每架飞机只选一条Lof的constraint
}

int Aircraft::_count = 0;

void Aircraft::print()
{
	//* time_t startT = _startT - TIMEDIFF;
	//* time_t endT = _endT - TIMEDIFF;
	time_t startT = _startT;
	time_t endT = _endT;
	cout <<"Tail " << _tail <<  endl;
	cout << _depStation->getName() << " " << ctime(&startT);
	cout << _arrStation->getName() << " " << ctime(&endT); 
}

void Aircraft::sortScheLegByDepTime()
{
	sort(_planLegList.begin(), _planLegList.end(), Leg::compareDepTime);
}

bool Aircraft::isPlanLegFeasible()
{
	// 检查最前和最后一个leg是否满足aircraft start/end time
	if (_planLegList.front()->getDepTime() < _startT)
		return false;
	if (_planLegList.back()->getArrTime() > _endT)
		return false;

	// 检查最前和最后一个leg是否满足aircraft start/end airport
	if (_planLegList.front()->getDepStation() != _depStation)
		return false;
	if (_planLegList.back()->getArrStation() != _arrStation)
		return false;

	Leg* thisLeg = NULL;
	Leg* nextLeg = NULL;

	// 检查airport match, time mismatch
	for (int i = 0; i < _planLegList.size() - 1; i++)
	{
		thisLeg = _planLegList[i];
		nextLeg = _planLegList[i+1];
		if (thisLeg->getArrStation() != nextLeg->getDepStation())
			return false;

		if (!thisLeg->isMaint() && !nextLeg->isMaint())		// flight, flight
		{
			if (thisLeg->getArrTime() + Util::turnTime > nextLeg->getDepTime())
				return false;
		}
		else// flight, maint; maint, flight; maint, maint
		{
			if (thisLeg->getArrTime() > nextLeg->getDepTime())
				return false;
		}
	}
	
	// 检查maintenance和aircraft是否match; dep time和arr time是否受到airport closure影响
	vector<pair<time_t, time_t>> depCloseList;
	vector<pair<time_t, time_t>> arrCloseList;
	for (int i = 0; i < _planLegList.size(); i++)
	{
		thisLeg = _planLegList[i];
		if (thisLeg->isMaint())
		{
			if (thisLeg->getAircraft() != this)
				return false;
		}
		else // flight check if affected by airport closure
		{
			depCloseList = thisLeg->getDepStation()->getCloseTimeList();
			for (int k = 0; k < depCloseList.size(); k++)
			{
				if (thisLeg->getDepTime() >= depCloseList[k].first &&
					thisLeg->getDepTime() < depCloseList[k].second)
					return false;
			}

			arrCloseList = thisLeg->getArrStation()->getCloseTimeList();
			for (int k = 0; k < arrCloseList.size(); k++)
			{
				if (thisLeg->getArrTime() >= arrCloseList[k].first &&
					thisLeg->getArrTime() < arrCloseList[k].second)
					return false;
			}
		}
	}

	return true;
}

/*省略了剩下的成员函数*/